libp2p/autotls/autotls.nim

```diff
@@ -0,0 +1,271 @@
+    keyAuthorization: Opt.none(string),
+    ipAddress: ipAddress,
+  )
+
+proc checkDNSRecords(
+    self: AutoTLSManager,
+    ip4Domain: string,
+    acmeChalDomain: string,
+    retries: int = DefaultDnsRetries,
+): Future[bool] {.async: (raises: [AutoTLSError, CancelledError]).} =
+  var txt: seq[string]
+  var ip4: seq[TransportAddress]
+
+  for _ in 0 .. retries:
+    txt = await self.dnsResolver.resolveTxt(acmeChalDomain)
+    try:
+      ip4 = await self.dnsResolver.resolveIp(ip4Domain, 0.Port)
+    except CatchableError as exc:
+      error "Failed to resolve IP", description = exc.msg # retry
+    if txt.len > 0 and self.keyAuthorization.isSome and
+        txt[0] == self.keyAuthorization.get() and ip4.len > 0:
+      return true
+    await sleepAsync(DefaultDnsRetryTime)
+
+  return false
+
+method issueCertificate(
+    self: AutoTLSManager
+): Future[void] {.base, async: (raises: [AutoTLSError, CancelledError]).} =
+  trace "Issuing new certificate"
+  let peerInfo = self.peerInfo.valueOr:
+    raise newException(AutoTLSError, "Cannot issue new certificate: peerInfo not set")
+
+  # generate autotls domain string: "*.{peerID}.libp2p.direct"
+  let base36PeerId = encodePeerId(peerInfo.peerId)
+  let baseDomain = base36PeerId & "." & AutoTLSDNSServer
+  let domain = "*." & baseDomain
+
+  trace "Requesting ACME challenge"
+  let (dns01Challenge, finalizeURL, orderURL) =
+    await self.acmeAccount.requestChallenge(@[domain])
+
+  self.keyAuthorization = Opt.some(
+    base64UrlEncode(
+      @(
+        sha256.digest(
+          (
+            dns01Challenge.getJSONField("token").getStr & "." &
+            thumbprint(self.acmeAccount.key)
+          ).toByteSeq
+        ).data
+      )
+    )
+  )
+
+  var strMultiaddresses: seq[string] = @[]
+  for ma in peerInfo.addrs:
+    strMultiaddresses.add($ma)
+
+  let payload = %*{"value": self.keyAuthorization.get(), "addresses": strMultiaddresses}
+  let registrationURL = "https://" & AutoTLSBroker & "/v1/_acme-challenge"
+
+  trace "Sending challenge to AutoTLS broker"
+  var response: HttpClientResponseRef
+  var bearerToken: string
+  if self.bearerToken.isSome:
+    (bearerToken, response) = await peerIdAuthSend(
+      registrationURL,
+      self.httpSession,
+      peerInfo,
+      payload,
+      self.rng,
+      bearerToken = self.bearerToken,
+    )
+  else:
+    # authenticate, send challenge and save bearerToken for future requests
+    (bearerToken, response) = await peerIdAuthSend(
+      registrationURL, self.httpSession, peerInfo, payload, self.rng
+    )
+    self.bearerToken = Opt.some(bearerToken)
+  if response.status != HttpOk:
+    raise newException(
+      AutoTLSError, "Failed to authenticate with AutoTLS Broker at " & AutoTLSBroker
+    )
+
+  # no need to do anything from this point forward if there are not public ip addresses on host
+  var hostPrimaryIP: IpAddress
+  try:
+    hostPrimaryIP = self.ipAddress.get(checkedGetPrimaryIPAddr())
```

Comment:
I wonder if this `get` gets lazily evaluated? because if not, maybe something like this can be used to avoid callign checkedGetPrimaryIPAddr if not needed:
```
let hostPrimaryIP: IpAddress = self.ipAddress.valueOr:
   try:
      let ip = checkedGetPrimaryIPAddr()
      if not isPublicIPV(ip):
         raise newException(AutoTLSError, "Host does not have a public IPv4 address")
      ip
   except GetPrimaryIPError as exc:
     raise newException(AutoTLSError, "Failed to get primary IP address for host", exc)
   except CatchableError as exc:
     raise newException(
      AutoTLSError, "Unexpected error while getting primary IP address for host", exc
    )
```